src/main/java/org/testng/internal/TestResult.java

```diff
@@ -320,4 +336,89 @@ public boolean wasRetried() {
+      if (isGlobalFailure(skippedConfig) || isRelated(skippedConfig)) {
+        //If there's a failure in @BeforeTest/@BeforeSuite/@BeforeClass
+        //then the reason is most often just one method.
+        skippedDueTo.add(skippedConfig.getMethod());
+        return Collections.unmodifiableList(skippedDueTo);
+      }
+      if (belongToSameGroup(skippedConfig)) {
+        //If its @BeforeGroups then there's a chance that there could be more than one
+        //method. So lets add everything.
+        skippedDueTo.add(skippedConfig.getMethod());
+      }
+    }
+    if (!skippedDueTo.isEmpty()) {
+      //If we found atleast one skipped due to reason, then its time to return back.
+      return Collections.unmodifiableList(skippedDueTo);
+    }
+    //Looks like we didnt have any configuration failures. So some upstream method perhaps failed.
+    List<String> upstreamMethods = Arrays.asList(m_method.getMethodsDependedUpon());
+    if (upstreamMethods.isEmpty()) {
+      //Maybe group dependencies exist ?
+      List<String> upstreamGroups = Arrays.asList(m_method.getGroupsDependedUpon());
+      if (upstreamGroups.isEmpty()) {
+        return Collections.emptyList();
+      }
+      List<ITestResult> allFailures = Lists.merge(m_context.getFailedTests().getAllResults(),
+          m_context.getFailedButWithinSuccessPercentageTests().getAllResults());
+      skippedDueTo = allFailures.stream()
+          .map(ITestResult::getMethod)
+          .filter(method -> {
+            List<String> currentMethodGroups = Arrays.asList(method.getGroups());
+            List<String> interection = Lists.intersection(upstreamGroups, currentMethodGroups);
+            return !interection.isEmpty();
+          }).collect(Collectors.toList());
+
+      return Collections.unmodifiableList(skippedDueTo);
+    }
+
+    //So we have dependsOnMethod failures
+    List<ITestResult> allfailures = Lists.merge(m_context.getFailedTests().getAllResults(),
+        m_context.getFailedButWithinSuccessPercentageTests().getAllResults());
+    skippedDueTo = allfailures.stream()
+        .map(ITestResult::getMethod)
+        .filter(method -> upstreamMethods.contains(method.getQualifiedName()))
+        .collect(Collectors.toList());
+    return Collections.unmodifiableList(skippedDueTo);
+  }
+
+  private static boolean isGlobalFailure(ITestResult result) {
+    ITestNGMethod m = result.getMethod();
+    return m.isBeforeTestConfiguration() || m.isBeforeSuiteConfiguration();
+  }
+
+  private boolean isRelated(ITestResult result) {
+    ITestNGMethod m = result.getMethod();
+    if (!m.isBeforeClassConfiguration() && !m.isBeforeMethodConfiguration()) {
+      return false;
+    }
+    Object current = this.getInstance();
+    Object thatObject = result.getInstance();
+    return current.getClass().isAssignableFrom(thatObject.getClass()) ||
+        thatObject.getClass().isAssignableFrom(current.getClass());
+  }
+
+  private boolean belongToSameGroup(ITestResult result) {

```

Comment:
Same here with `private static boolean belongToSameGroup(ITestResult result, String[] groups)`

Reply:
Same case here as well. Here we are basically checking if two `ITestResult` objects belong to the same set of groups. That is why its an instance method.